File: algorithms/179.Largest_Number/Largest_Number.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    static bool cmp(string a, string b) {
        return (a + b) > (b + a);
    }

public:
    string largestNumber(vector<int> &num) {
        vector<string> str;
        for (auto &i : num)
            str.push_back(to_string(i));

        sort(str.begin(), str.end(), cmp);

        string ans;
        if (str[0] == "0")
            return "0";
        for (auto &i : str)
            ans += i;

        return ans;
    }
};
class Solution {
    public:

        //compare function
        static bool comp (string& s1, string& s2) { return s1+s2 > s2+s1; }

        string largestNumber(vector<int> &num) {

            //convert int to string
            vector<string> v;
            for (int i=0; i<num.size(); i++) {
                v.push_back(to_string(num[i]));
            }

            //sort the string
            sort(v.begin(), v.end(), comp);

            //generate the result
            string result;
            for(int i=0; i<v.size(); i++ ){
                result+=v[i];
            }

            //special case: start with zero (e.g. [0, 0])
            if (result[0]=='0') return "0";

            return result;
        }
};

int main() {
    int n, t;
    Solution s;
    cin >> n;
    vector<int> num(n);
    for (auto &i : num)
        cin >> i;
    cout << s.largestNumber(num) << endl;
    return 0;
}

File: algorithms/179.Largest_Number/Largest_Number.java
import java.util.Comparator;
import java.util.Scanner;

public class Solution {

    public String largestNumber(int[] num) {
        String[] str = new String[num.length];
        for (int i = 0; i < num.length; ++i) {
            str[i] = num[i] + "";
        }

        Comparator<String> comp = new Comparator<String>() {
            @Override
            public int compare(String o1, String o2) {
                return (o2 + o1).compareTo(o1 + o2);    // reverse
            }
        };

        Arrays.sort(str, comp);
        if (str[0].equals("0")) {
            return "0";
        }

        String ans = new String();
        for (String s : str)
            ans += s;

        return ans;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        int[] input = {0, 0, 0};
        System.out.println(s.largestNumber(input));
    }
}
class NumbersComparator implements Comparator<String> {
	@Override
	public int compare(String s1, String s2) {
		return (s2 + s1).compareTo(s1 + s2);
	}
}

public class Solution {
    
    public String largestNumber(int[] nums) {
        String[] strs = new String[nums.length];
        for (int i = 0; i < nums.length; i++) {
            strs[i] = Integer.toString(nums[i]);
        }
        Arrays.sort(strs, new NumbersComparator());
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < strs.length; i++) {
            sb.append(strs[i]);
        }
        String result = sb.toString();
        int index = 0;
        while (index < result.length() && result.charAt(index) == '0') {
            index++;
        }
        if (index == result.length()) {
            return "0";
        }
        return result.substring(index);
    }
}